### PR TITLE
Add Redis to collaboration service

### PR DESCRIPTION
**Pull request by cadencjk**
- Add Redis to collaboration service, so that users can recover a match and get the synchronised data
- Add execution time whenever code is executed
- Store execution output to history
- Fix bugs
- Update documentations

### CODE REVIEW

File: History/src/HistoryController.js [hunk 1]
```diff
@@ -4,10 +4,10 @@ const { Status } = require('./constants');
 const logger = require('./Log');
 
 exports.addAttempt = async (req, res) => {
-  const { userId, questionId, code, language } = req.body;
+  const { userId, questionId, code, language, result } = req.body;
   try {
-    await historyService.addAttempt(userId, questionId, code, language);
-    logger.logSuccess(`Added attempt for user ${userId}`);
```
Comment (JrmCkh) [inline History/src/HistoryController.js:10]: Log message was updated to this for clarity

File: Collaboration/src/CollaborationServer.js [hunk 2]
```diff
@@ -3,6 +3,8 @@ const socketIo = require('socket.io');
 const env = require('./loadEnvironment');
 const logger = require('./Log');
 const { Event } = require('./constants');
+const { getBoilerplate } = require('./utils/helpers');
+const { RedisMemory } = require('./RedisMemory');
```
Comment (JrmCkh) [inline Collaboration/src/CollaborationServer.js:8]: Perhaps can change the ordering to 
1. ./Redis
2. ./utils
3. ./constants
4. ./load
5. ./Log

File: Collaboration/src/CollaborationServer.js [hunk 3]
```diff
@@ -117,12 +157,27 @@ try {
...
       io.to(room).emit(Event.Communication.CHAT_RECEIVE, message);
+
+      redisMemory.handleChatMessage(room, message);
+    });
+
+    // Handle code reset event
+    socket.on(Event.Code.RESET, (data) => {
+      const room = data.room;
+      const language = data.language;
+
```
Comment (JrmCkh) [inline Collaboration/src/CollaborationServer.js:168]: Perhaps can remove this line to collapse the const together

File: docs/ApiEndpoints.md [hunk 4]
```diff
@@ -17,41 +17,41 @@
...
-| PUT    | `/user/display-name`         | Update user's display name                      | `id` <br> `username`                                                  | JWT token       | User            |
-| PUT    | `/user/change-password`      | Change user's password                          | `id` <br> `currentPassword` <br> `newPassword` <br> `confirmPassword` | JWT token       | User            |
-| PUT    | `/user/programming-language` | Update user's preferred programming language    | `id` <br> `programmingLanguage`                                       | JWT token       | User            |
-| PUT    | `/user/complexity`           | Update user's preferred question complexity     | `id` <br> `complexity`                                                | JWT token       | User            |
+| Method | API Endpoints                     | Purpose                                                                           | Parameters <br> (JSON format)                                         | Header Contains | [Roles](#roles) |
```
Comment (JrmCkh) [inline docs/ApiEndpoints.md:20]: Did you try to 'format document' this page?

File: docs/Containerization.md [hunk 5]
```diff
@@ -4,13 +4,14 @@ Docker is used to containerize our various micro-services
 which serves our combined application.
 
 We have various micro-services including:
-* MySQL Database
+* Redis Service
```
Comment (JrmCkh) [inline docs/Containerization.md:7]: I don't think Redis itself is considered a micro-service, while it is true that it is a service that we use.

File: docs/LocalHosting.md [hunk 6]
```diff
@@ -41,9 +44,6 @@ npm run install-all
 
 # Start Application Locally
 
-Ensure that MySQL Server service has started.
-(How to check? Click [here](LocalMySqlSetup.md#start-mysql-service).)
-
 Start all services using the following commands:
```
Comment (JrmCkh) [inline docs/LocalHosting.md:47]: Perhaps there a need to specify to 'ensure that xxx service has started' for RabbitMQ/Redis/Ubuntu?

File: frontend/src/components/Collaboration/CodeEditor.js [hunk 7]
```diff
@@ -1,4 +1,5 @@
 import React, { useState, useEffect, useRef } from 'react';
+import Spinner from 'react-bootstrap/Spinner';
```
Comment (JrmCkh) [inline frontend/src/components/Collaboration/CodeEditor.js:2]: Wrong import perhaps? I don't see any usage of Spinner + I think we have a Spinner component within our Component folder.

**Review by JrmCkh — CHANGES_REQUESTED** [review]
Some minor changes required.

File: docs/ApiEndpoints.md (hunk 4 above)
Comment (cadencjk) [reply docs/ApiEndpoints.md:20]: image

Already formatted on my page, might be a github display thing

File: docs/LocalRedisSetup.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  - Port should be `6379`
+
+## Start Redis Service
+
+For Windows OS and Mac OS,
+1. Start your Ubuntu terminal
+2. Type `redis-server` into the terminal
+3. Leave the terminal running
+
+[Go to README](../README.md)
```
Comment (lhy-hoyin) [inline docs/LocalRedisSetup.md:21]: Add a `---` above the README link, to add a horizontal line

**Review by JrmCkh — APPROVED** [review]
LGTM

**Pull request merged**